### DISCUSSION TITLE
Why is log.With(fields) slower than log.Info(msg, fields)

### DISCUSSION THREAD

**Discussion by dmdok** [uber-go/zap#1112] [Q&A] [answered]
consoleEncoder uses jsonEncoder. jsonEncoder writes fields immediately in the buffer. Then it copies the buffer when log.With is called. This behaviour is very unfortunate if logger is never called. And that is quite possible if logging is used mostly for errors.
Is there a way to make log.With lazy? For example, when buffer is cached on logger first call only.

**Comment by abhinav** [comment, marked as answer]
Hello! Yes, indeed log.With is eager at serializing because it's meant for loggers that are meant to be 

```
logger2 := logger.With(zap.String("foo", "bar"))
// serializes the field immediately so that
// the following does not pay to serialize that field twice.

logger2.Info("stuff happened")  // includes {"foo": "bar"} in the output
logger2.Info("more stuff happened")  // same
```

logger.With shares the serialized form of those fields between all uses of the logger.

For fields that are not meant to be re-used across loggers, we recommend this pattern:

```
logger.Info("stuff happened", zap.String("foo", "bar"))
```

This will serialize those fields only if that log message is actually posted.

We'll try to add something about this to the documentation, but to summarize:

```
logger.With([fields]).Info("foo")  // bad
logger.Info("foo", [fields])  // good
```

We're actually trying to build a small linter to catch these cases; we're hoping to open source it once it's ready.

---

Hope this helps! Closing because answered. Feel free to re-open or create a new ticket.
